mirrord/agent/src/vpn.rs

```diff
@@ -0,0 +1,351 @@
+    let local_address = SocketAddr::new(temporary_socket.local_addr()?.ip(), 0);
+    let raw_local_address = SockaddrStorage::from(local_address);
+
+    // Try to find an interface that matches the local ip we have.
+    let usable_interface = nix::ifaddrs::getifaddrs()?
+        .find(|iface| {
+            iface
+                .address
+                .map(|addr| addr == raw_local_address)
+                .unwrap_or(false)
+        })
+        .ok_or_else(|| AgentError::VpnError("usable_interface".to_owned()))?;
+
+    let ip = usable_interface
+        .address
+        .ok_or_else(|| AgentError::VpnError("usable_interface".to_owned()))?
+        .as_sockaddr_in()
+        .ok_or_else(|| AgentError::VpnError("usable_interface".to_owned()))?
+        .ip()
+        .into();
+    let net_mask = usable_interface
+        .netmask
+        .ok_or_else(|| AgentError::VpnError("usable_interface".to_owned()))?
+        .as_sockaddr_in()
+        .ok_or_else(|| AgentError::VpnError("usable_interface".to_owned()))?
+        .ip()
+        .into();
+    // extracting gateway is more difficult, ugly patch for now.
+    let temp_gateway = usable_interface
+        .address
+        .ok_or_else(|| AgentError::VpnError("usable_interface".to_owned()))?
+        .as_sockaddr_in()
+        .ok_or_else(|| AgentError::VpnError("usable_interface".to_owned()))?
+        .ip()
+        .octets();
+
+    let gateway = IpAddr::V4(Ipv4Addr::new(
+        temp_gateway[0],
+        temp_gateway[1],
+        temp_gateway[2],
+        1,
+    ));
+
+    Ok((ip, net_mask, gateway))
+}
+
+/// Handles outgoing connections for one client (layer).
+struct VpnTask {
+    pid: Option<u64>,
+    layer_rx: Receiver<ClientVpn>,
+    daemon_tx: Sender<ServerVpn>,
+    socket: Option<AsyncRawSocket>,
+}
+
+impl fmt::Debug for VpnTask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("VpnTask").field("pid", &self.pid).finish()
+    }
+}
+
+fn interface_index_to_sock_addr(index: i32) -> Result<SockAddr> {
+    let mut addr_storage: libc::sockaddr_storage = unsafe { std::mem::zeroed() };
+    let len = std::mem::size_of::<libc::sockaddr_ll>() as libc::socklen_t;
+    let macs = procfs::net::arp().map_err(|err| AgentError::VpnError(err.to_string()))?;
+    tracing::debug!(?macs, "arp entries");
+
+    let hw_addr = macs
+        .into_iter()
+        .find_map(|entry| entry.hw_address)
+        .ok_or_else(|| AgentError::VpnError("no entry with hw_address".to_owned()))?;
+
+    unsafe {
+        let sock_addr = std::ptr::addr_of_mut!(addr_storage) as *mut libc::sockaddr_ll;
+        (*sock_addr).sll_family = libc::AF_PACKET as u16;
+        (*sock_addr).sll_protocol = (libc::ETH_P_IP as u16).to_be();
+        (*sock_addr).sll_ifindex = index;
+        (*sock_addr).sll_halen = 6;
+        (*sock_addr).sll_addr = [
+            hw_addr[0], hw_addr[1], hw_addr[2], hw_addr[3], hw_addr[4], hw_addr[5], 0, 0,
+        ];
+    }
```

Comment:
Could probably be nicer with [LinkAddr](https://docs.rs/nix/latest/nix/sys/socket/struct.LinkAddr.html)